### PR TITLE
ZDC/RPD QIE10 Digis DQM online

### PR DESCRIPTION
**Pull request by BetterWang**
Online DQM monitoring for ZDC/RPD detector on QIE10 Digis.

### CODE REVIEW

**Comment by BetterWang** [comment]
@mandrenguyen HI software contact person.

**Comment by andrius-k** [comment]
please test

**Comment by abdoulline** [comment]
@DryRun  @deguio 
Just to draw your attention

**Comment by deguio** [comment]
hi @BetterWang 
why don't you update the existing ZDCTask instead of creating a new one?

**Comment by BetterWang** [comment]
Hi @deguio , the old one was for QIE8.

**Comment by andrius-k** [comment]
Hi Quan,

Is it really necessary to use null-terminated strings (`.c_str()`)? The bot seems to be complaining about that in code checks.

**Comment by deguio** [comment]
@BetterWang right..

**Comment by BetterWang** [comment]
right...

**Comment by deguio** [comment]
ok. let me rephrase. do you really need to duplicate an existing plugin instead of adapting the existing one to use the new QIE10 collection?

**Comment by andrius-k** [comment]
please test

**Comment by abdoulline** [comment]
That was actually my (unasked) question yesterday...
  ZDC QIE8 was over since 2016. And existing code contains QIE10
  preparations. Why keeping QIE8-related parts?

**Comment by BetterWang** [comment]
Hi @abdoulline @deguio , we can definitely replace the QIE8 module, if that's what you prefer.

**Comment by andrius-k** [comment]
Hi,

Will the existing plugin be updated to use the QIE10 collection? In such case this PR could be closed.

**Comment by BetterWang** [comment]
Hi @andrius-k , I don't think the existing plugin can be upgraded to QIE10 easily. ZDC will use the QIE card without winchester connectors, which means, no calibration constants for conversion  to charge. This DQM module is only monitoring the DIGIs.

**Comment by andrius-k** [comment]
+1
Ok, thanks for the explanation @BetterWang.

**Comment by abdoulline** [comment]
@BetterWang, @deguio 

I'd suggest that at some point, but *not now*,  to remove old QIE8-related stuff (which is not and will not be used anywhere).

**Comment by fabiocos** [comment]
@abdoulline @deguio so could you please finally clarify if this PR is ok for the HCAL DPG? 
@andrius-k @BetterWang this PR is relevant for online DQM, but running also in offline sequences, am I correct?

File: DQM/HcalTasks/interface/ZDCQIE10Task.h [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+	protected:
+		void analyze(edm::Event const&, edm::EventSetup const&) override;
+
+		//	tags
+		edm::InputTag	_tagQIE10;
+		edm::EDGetTokenT<QIE10DigiCollection> _tokQIE10;
+
+		//	hcaldqm::Containers
+		std::map<std::string,MonitorElement*>   _cADC_EChannel;
```
Comment (davidlange6) [inline DQM/HcalTasks/interface/ZDCQIE10Task.h]: hi- i had three comments on this PR
1) your data is much better suited for something like a 
std::array< std::array< std::array<MonitorElement *, 6>, 2 > 4 > _cADC_EChannel 
to avoid the endless conversions to and from strings (per hit!)

2) avoid sprintfs

3) avoid unneeded calls to
ib.setCurrentFolder("Hcal/ZDCQIE10Task/Shape_perChannel");

Comment (BetterWang) [reply DQM/HcalTasks/interface/ZDCQIE10Task.h]: Hi @davidlange6 , just made another commit.
1. This is a really sparse array, so I prefer map.
2. sprintf -> string
3. Done.

**Comment by BetterWang** [comment]
Hi @fabiocos , this PR is meant for both online and offline.

**Comment by deguio** [comment]
hi @BetterWang 
as it is I don't think that the new analyzer runs anywhere. I don't see it included in any sequence neither for online or offline.

@fabiocos my comment was about the old `zdcTask` which is orphan at this point. it can be removed as salavat mentioned.

for the online part @DryRun should be able to give inputs about the inclusion of the new `zdcTask` into
https://github.com/cms-sw/cmssw/blob/master/DQM/Integration/python/clients/hcal_dqm_sourceclient-live_cfg.py

File: DQM/HcalTasks/interface/ZDCQIE10Task.h (hunk 1 above)
Comment (DryRun) [reply DQM/HcalTasks/interface/ZDCQIE10Task.h]: I second comment (1) -- perhaps @davidlange6 's suggest would result in a sparsely populated array (though it's just pointers, so maybe not a big deal), but it wouldn't be hard to come up with a different solution that avoids performing several string conversions and comparisons for every hit. For example, defining separate maps for each of (EM, HAD, LUM, RPD) * (plus, minus), each a `std::map <int, MonitorElement*>` or even just an array of the desired size, or making an enum like`ZDCSubdet_t = {EMP, HADP, LUMP, RPDP, EMM, HADM, LUMM, RPDM};` and storing things in `std::map<ZDCSubdet_t, std::vector<MonitorElement*> >`.

**Comment by BetterWang** [comment]
hi @deguio , Yes, once this PR is merged, I will make another PR to include this to online and offline DQM sequence. One more thing, this is meant for PbPb run using 10_3_X. So far, the ZDC/RPD emap has not been integrated to the GT yet. So, there is no point to put them in the sequence now.

File: DQM/HcalTasks/interface/ZDCQIE10Task.h (hunk 1 above)
Comment (BetterWang) [reply DQM/HcalTasks/interface/ZDCQIE10Task.h]: Hi @DryRun @davidlange6 , since nobody likes short string optimization, I switched to detector rawId as the key in the new commit.

**Comment by fabiocos** [comment]
please test

**Comment by fabiocos** [comment]
@andrius-k @jfernan2 could you please review again this PR?

**Comment by andrius-k** [comment]
+1

**Comment by fabiocos** [comment]
+1
